### DISCUSSION TITLE
Move custom central server drop-down to connection dialog

### DISCUSSION THREAD

**Discussion by jp8** [jamulussoftware/jamulus#1869] [UI Related]
Hi Volker - would you consider moving this setting to the connection dialog, instead of the settings dialog?
Personally, any time I need to change to a different central server, I am already on the connection dialog.
It could even be integrated with the drop-down that lists the 'fixed' central servers....

**Comment by corrados** [comment]
The list selection combo box in the connection dialog only contains fixed entries. A combo box either has fixed entries or you can edit the content. A mixture is usually not a good thing. We would require additional GUI controls.

One problem is that there are a lot of tutorials which describe to set the server list to "Custom". If we change this, all these tutorials are no longer valid.

BTW: I have implemented a direct update of the list already. I.e., if you have the connection dialog and the settings dialog both open and you have the list selection set to Custom; if you then change the Custom Central Server Address, the server list is updated immediately.

**Comment by jp8** [comment]
I'm usually working on a 1024x768 screen, so would prefer if i didn't have to find settings on 2 different dialogs.

Once a user has entered a new custom server, there doesn't seem to be a way to remove it.  Therefore they are all 'fixed' entries.  Any use case for removing previously entered custom values probably extends to removing the 'fixed' initial entries too.

I think most of the tutorials that people are using are 'living' documents, and are likely to be edited to include changes and enhancements in Jamulus.  My own experience is no matter how clear I make the tutorial, many users end up with the custom central server value in the wrong field anyway.  They put it in the Server Address field on the connection dialog.

In fact, I would welcome a single 'custom server' entry.  After entering a custom server, the Jamulus client would detect the type of server

Simple server -> just connect to it
Central server ->  add it to the drop-down, auto-select it, and present the user with the list of attached servers.

**Comment by LauraMardones** [comment]
Jurriaan and I spent the better part of an hour trying to get a WHOLE BUNCH of people to understand the concept of a custom central server. This feature is one of the most important and most counter-intuitive functions of Jamulus. 

_(In the end, we just gave up, and moved the whole rehearsal into the central server, where all these people were stuck - Mohammed vs Mountain.)_

So consider the User Journey here. If I'm at the point where I understand I need to CONNECT to a server, then all the interaction related to connecting to servers should be found in that one Dialogue Window called "Connection setup". 

The case in point is this: a lot of users incorrectly type up the custom central server address in the "Server address" field in Connection setup. Even if they afterwards DO find the Settings window, and input the custom server address in the correct field, the same address is STILL remembered in the "Server address" of the Connection Setup window. The hapless, user, having filled out the same information in several places, has no idea what goes where now, and starts pushing the Connect button over and over again - always landing directly inside the central server, never managing to find the dropdown list and the servers in the custom central server...

There has to be a better way to poka-yoke this step.

**Comment by kezzy1966** [comment]
Laura, I think your suggestion would be nice thing to have and I really understand the difficulties of managing a choir with Jamulus. This is an excellent product but it was not specifically designed/ written for large choirs. It has been adapted to accommodate them. 

At this point of its development, everyone using it needs to take time to learn how to drive it. So in your example, how to change the CCSA in 'Settings'. 

Personally I would like to see a comprehensive discussion of all changes required to effectively manage large choir. I have raised some points already in GIT HUB specifically for management of large groups.  

Perhaps there could be a specific section on jamulus.io for 'running a choir'. That would be a user friend

**Comment by wschell** [comment]
My english is not so good, maybe I have missed something in the previous posts. 
What about this workflow: If I enter a custom central-server-adress in the connection dialog-field, jamulus will detect this, shows an optional ([ ] don't show this message again) dialog box, informing the user and if the user accepts, this server would be added to the custom-server-list in the settings-dialog and selected. The server type changes to custom. 
Additional, in the adress-combolist, custom servers could be marked i.e. with a "*" or in bold or italic or something like that.
In the future there would be no need for the custom-server-list in the settings dialog.
What do you think about it?

Edit: This would elminate the difference between custom server and custom central server - which from a users view is more technical. The user just want to connect.

**Comment by jugo6** [comment]
I was with Laura on the big Xmas Choir. I think we lost a couple of people who tried figuring out the central server. Of course there is a certain amount to blame on them. We gave a detailed instruction what to do where, but even the highest educated people just do something (perhaps especially the academics... :-) ) . Another handicap was, that we had people from 10 countries, and instructions in only English. 

Part 2 of the problem Laura said might even be bigger: by clicking connect, you will always go to the server adress (IP), and doubleclicking is a bit tricky because servers are moving in the list sometimes (ping). 

Might it be an idea (though radical) to move all those things to below the list:
(or above) so that related items end next to each other?

 List                    Custom Central Server Adress
[dropdown]       [fill in / drop down]
______
list
list
____
filter  [fill in ]    [x]  show all musicians
Direct Server Adress  [fill in / drop down]

**Comment by jugo6** [comment]
> Perhaps there could be a specific section on jamulus.io for 'running a choir'. That would be a user friend

We do have a facebook group Jamulus Choral Community for choir related things

**Comment by corrados** [comment]
> Jurriaan and I spent the better part of an hour trying to get a WHOLE BUNCH of people to understand the concept of a custom central server.

I guess we are talking about non-techies here. So any interaction with any GUI is critical. Moving a GUI control from one place to another does not solve the issue.

There are different solutions for the problem possible. E.g. you could create a Windows shortcut which loads a pre-configured ini-file (i.e. the shortcut calls Jamulus with the --inifile command line flag). The shortcut together with the pre-defined ini file should then sent to them via, e.g., email. A simple click on the shortcut will then configure Jamulus correctly. In the inifile even the selected server list is stored so you could set the correct Custom Central Server Address and also set the Server List to Custom. So the user does not need to configure anything.

If even the double click on the server list is too complicated, then you could additionally use the -c command line argument where Jamulus automatically connects to a server address on startup. I guess the choral servers are cloud servers with fixed IP addresses so the -c command line argument should work just fine.

You could create your own Windows Jamulus installer by putting the ini-file and shortcut in the installer. A long time ago someone just did this successfully (before pandemic, actually).

**Comment by houbsta** [comment]
Here's just an idea on how there could be a mirror of the current settings within the connection window also: 
![image](https://user-images.githubusercontent.com/69164084/103925009-2f06bb00-5117-11eb-8a1e-5ea6a2fd2223.png)

I think there is a lack of clarity because "custom server" is a different thing from "Server Name" or "Server Address" hence the addition of a label to try to clarify a little.

**Comment by houbsta** [comment]
Also I just thought: the "Server Address" dropdown should contain history of all servers connected to, not just those that you type / paste in yourself. That would be more coherent. A single click on a server in the list should populate the "server address" field. That will eliminate errors and confusion.

**Comment by LauraMardones** [comment]
I like Houbstas proposal. It doesn't solve the big problem of the techiness of Jamulus, but at least it's slightly more intuitive.

**Comment by LauraMardones** [comment]
> I guess we are talking about non-techies here. So any interaction with any GUI is critical. Moving a GUI control from one place to another does not solve the issue.

Fair point.

> There are different solutions for the problem possible. E.g. you could create a Windows shortcut which loads a pre-configured ini-file (i.e. the shortcut calls Jamulus with the --inifile command line flag). The shortcut together with the pre-defined ini file should then sent to them via, e.g., email. A simple click on the shortcut will then configure Jamulus correctly. In the inifile even the selected server list is stored so you could set the correct Custom Central Server Address and also set the Server List to Custom. So the user does not need to configure anything.

That's a really intriguing solution! I will try that for my next Jamulus event. <3

**Comment by hoffie** [comment]
@dcorson-ticino-com Do you know if this is something which is covered by the UI redesign? Should we keep this issue open?

**Comment by dcorson-ticino-com** [comment]
I see two issues here: choosing a central server and defining favorites.
I started making favorites once, but it is not simple and my understanding of Jamulus at the time was more limited.
Maybe I should look at it again.  My Ansatz was to have a favorites list shown on a second tab of the connect window, last used at the top.  Servers were added by clicking an Add to Favorites button when connected to the server.  Thus it works with private servers too.
But, as I said I have not continued with this and have not changed anything with central server definition either. 
If someone makes a definition I would look at it.  This use with big choirs is, however, outside my experience area.

**Comment by houbsta** [comment]
I have a fundamental issue with defining where a list comes from (on the settings screen, "custom central server address") and where you actually see the list (on the connection screen).

There are a number of ways this problem can be solved. My proposition was to make the minimum UI change possible and retain logic of the existing application, except that a definition now moves from the settings screen to the connection screen.

In my mind, it's far more coherent to add the choice of central server to the connection screen. It's not a hardware or network setting like the main settings screen, it's a custom parameter that defines what you see in the connection screen.

As a user experience professional the only true response I can give is to test with users. However I do indeed have some directly observed evidence, accrued over nearly a year of supporting Jamulus users, that the current place for setting a custom server is a source of some confusion. It does not appear logical to people to change to "settings" to then come back to the connection screen and see what has changed.

So I disagree with Volker's original comment "Moving a GUI control from one place to another does not solve the issue.". Putting a GUI control in the right place indeed is a solution. 

The "quick connect" server address field should populate with a selected server from the custom list too, so that the list of server addresses be ALL servers you have connected to, not just the ones you've added manually.

**Comment by LauraMardones** [comment]
This has not been fixed. Caused problem for new user yesterday. Issue
should remain open.

Den sön 28 mars 2021 13:04Christian Hoffmann ***@***.***>
skrev:

**Reply by hoffie** [reply]
> This has not been fixed. Caused problem for new user yesterday. Issue should remain open.

Yes, I'm aware. I was just wondering if this issue was already worked on as part of a larger refactoring. Seems like it wasn't, which is why I'd kept it open.

Personally I fully agree with @houbsta.

The current interface design is non-intuitive in this regard and it should be rather straight forward to change this.

**Comment by gilgongo** [comment]
This seems to need to some more discussion on how or whether this should be implemented. So I'm moving it to discussions until such time as we can get a firm spec for either an issue to be worked on, or a PR.

**Reply by hoffie** [reply]
> This seems to need to some more discussion on how or whether this should be implemented.

I'm wondering how we would reach that state, though?

I agree that the original issue was rather convoluted and contained lots of comments. However, the last comments included very specific ideas how the issue could be solved (graphically). The only thing which was missing was someone to implement it, wasn't it?

By moving this (and similar) issues to discussions, I fear that
- we may lose track of things and may therefore discard valuable user input (which we struggle to acquire in other places)
- we may not increase chances that such a request is approved (by getting visibility by devs) as Discussions are higher volume and not every dev reads them.

**Reply by gilgongo** [reply]
> The only thing which was missing was someone to implement it, wasn't it?

In which case that sounds like the perfect point at which to raise an issue or PR that was born out of this discussion :-)  I wasn't able to tell what that was though.

Things in the issues list need to be actionable and agreed specs for work, or we go back to the state we were in about 18 months ago where the list of issues bore little resemblance to what was actually being considered for action. The only way you could tell was to read all the conversations. 

The alternative is to have a convention where the "ticket" (the first post in the issue thread) is always kept up to date to describe  modifications made to it in the discussion under it. But either people don't want (or can't?) do that for various reasons, so better to have the convention of raising an issue (or a PR) from a Discussion once a firm decision is made.

> By moving this (and similar) issues to discussions, I fear that

Those are exactly my fears too. But about the issues list.

**Reply by dcorson-ticino-com** [reply]
I'll do a PR for this.
Something similar to Houbsta's suggestion.
EDIT I think I will leave the terminology Central Server in the code, otherwise there are things to change all over in the code. OK?

**Reply by gilgongo** [reply]
Don't forget to link to the relevant context in this discussion that has given rise to the PR.

**Reply by jujudusud** [reply]
@gilgongo -> Those changes will also require to take a look to the documentation and the screenshots if applied.

**Reply by jp8** [reply]
Houbsta's screenshots are definitely a step in the right direction.

To go even further though, I don't personally see any need for two boxes to enter server addresses.  The users don't understand the difference, and will use the wrong box approximately half the time.

And I don't see any real reason for the terminology of 'custom' central server.  There is nothing 'custom' about it.  It's a central server just like the others.  Why not just add the user's extra central servers directly to the list at the top, alongside the genre-based central servers.

**Reply by houbsta** [reply]
Decisions need to be made carefully on changing nomenclature. Whether the existing term is correct has to be weighed against whether existing documentation, how-to guides and tutorial videos are likely to be updated in a timely manner. People are used to the way certain things work. Clarifying gradually with incremental change can be easier to adapt to than wholesale interface change.

My changes were made from an organic improvement point of view. Reworking the entire thinking was not in the scope of the screen design. Directory servers need to be distinct from individual servers - that's the main problem I was trying to solve. The user pain point - observed in research on many occasions by several different choir masters in particular - is that users enter "custom central/directory server" addresses into a box which does not serve that purpose at all.

**Reply by jp8** [reply]
Houbsta is right - the initial step of just moving the existing configurations to be side-by-side on the connection dialog is already a big improvement.

**Reply by gilgongo** [reply]
@jp8 and @houbsta These issues are bring discussed as part of this PR now:  https://github.com/jamulussoftware/jamulus/pull/1884

**Reply by jp8** [reply]
I've added an alternative pull request for this: #1894
